### ISSUE TITLE
Question: Best practices for some real-world scenarios.

### ISSUE THREAD

**Issue by zdavison** [realm/realm-cocoa#2218]
We use Realm here, and we're so far fairly happy with it, apart from a few snags we have mostly worked around (No null values in the DB? We use sentinel values. No KVO? We rolled our own Swift Observable that powers RAC), regardless, so far it is a nicer experience than Core Data, and being able to view the source and figure out the innards of what used to be a black box is important to us.

However, one thing we have issues with is how to handle thread safety and uniquing. Our original (current) solution was to never pass around actual Realm objects, and only pass around their primary keys, then, whenever an object needs to do anything in Realm, we pull it out of the DB. This works fine because we put an in-memory Realm in front of our actual on disk realm, so we incur no performance issues here.
However, this is sort of messy and feels like a code smell to me, like a giant band aid on a problem, realistically, what we want is the ability to create an immutable, thread safe **view** of a Realm model, and pass that around at our discretion, thus resolving any synchronization issues, and hence any thread safety issues. This model has a reference to its 'true' primary key, so we can `commit` it via Realm whenever we want.
Obviously though, this doesn't exist, so we're considering building it. We're wondering if this is worth the time, and if it makes sense, obviously Realm thread safety is an issue that pops up here frequently so is presumably on your minds.

Our core usage of Realm is as a cache sitting between us and the server. We don't really use it for anything else and the server is considered the 100% source of truth always, we only perform local mutation to keep local data in sync. One of the things we like about Realm is the ability to `update` existing Realm objects, hence uniquing them on a primary key, but the issue here is that our usage causes us to sometimes load partial objects from the server, that then **overwrite** objects which contain more data (eg: a partial object from the server with 2 populated fields can overwrite an object with all 20 fields populated). Again, we're unsure how to solve this, we are considering selectively storing things in realm, and rather than reaching in to Realm to get objects by ID, using the previously mentioned immutable, safe `views` of Realm objects and passing those around .This would mean we could selectively choose to not add things to Realm.

Basically, this is another query around the core problem of passing Realm objects between threads **before they have been added to a Realm**. 

Thanks for your time and please continue developing a great product.

**Comment by segiddins** [comment]
I agree that passing around primary keys is far from optimal, which is why we're working on a native way to have results delivered onto another queue. You can follow https://github.com/realm/realm-cocoa/issues/1097 for updates (I've been working on that issue for the past couple of weeks).

**Comment by segiddins** [comment]
@zdavison since I haven't heard back from you, I'm going to close this issue for now, but feel free to reopen it if you have any more questions!
